contrib/platform/src/com/sun/jna/platform/win32/WinGDI.java

```diff
@@ -28,18 +28,101 @@
 import com.sun.jna.Structure;
 import com.sun.jna.Structure.FieldOrder;
 import com.sun.jna.platform.win32.WinNT.HANDLE;
-import com.sun.jna.platform.win32.WinDef.HBITMAP;
-import com.sun.jna.platform.win32.WinDef.RECT;
+import com.sun.jna.Union;
+
+import static com.sun.jna.platform.win32.WinDef.*;
 
 /**
  * Ported from WinGDI.h.
  * Microsoft Windows SDK 6.0A.
  * @author dblock[at]dblock.org
  * @author Andreas "PAX" L&uuml;ck, onkelpax-git[at]yahoo.de
+ * @author SSHTOOLS Limited, support@sshtools.com
  */
 public interface WinGDI {
     int RDH_RECTANGLES = 1;
 
+    @FieldOrder({ "dmDeviceName", "dmSpecVersion", "dmDriverVersion", "dmSize", "dmDriverExtra", "dmFields", "dmUnion1", "dmColor",
+            "dmDuplex", "dmYResolution", "dmTTOption", "dmCollate", "dmFormName", "dmLogPixels", "dmBitsPerPel", "dmPelsWidth",
+            "dmPelsHeight", "dummyunionname2", "dmDisplayFrequency", "dmICMMethod", "dmICMIntent", "dmMediaType", "dmDitherType",
+            "dmReserved1", "dmReserved2", "dmPanningWidth", "dmPanningHeight" })
+
+    public static class DEVMODE extends Structure {
```

Comment:
Reopening this comment, since the previous one at this code marking ended up talking about fixing the pointer and then reading a wide string properly.

Quoting @dbwiddis:

> Unions are fine, but generally you should try to add code in the constructor that detects which union member needs to be read, and then reads it. That looks a little bit complex here... do you read fields and then based on which bits are set choose which union member applies?

@dbwiddis Is the current structure OK?  My structure is a copy from a project called "[damage](https://github.com/sshtools/damage/blob/701199ca45fb5994d778402ec497dd58b732fed2/src/main/java/org/x/ExWinGDI.java#L51)" and it seems to mirror the [Microsoft definition exactly](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-devmodea), take note of the words `DUMMY...` in the definition.

Reply:
> Is the current structure OK? 

I'd quibble over style; I'd prefer to see all the structure fields together at the top, and the nested class definitions below them.  Right now it's hard to see which elements are in the union.

Yes, it's fine to match the Windows API names.

There is one missing piece, though, and I'm not clear how the structure works.  In the native, all fields of the union are essentially "active" at any given time.  In JNA, we only "read" the member of the union which is declared to be the correct type.  For most unions this information is given in one of the other fields.

In this structure it looks like one of the fields is a bitmask indicating which of several union fields may be active.  So you might have to do some bit manipulation there, or perhaps just read all three possibilities, doing something like this in the `DEVMODE` structure:
```java
@Override
public void read() {
    super.read();
    dmUnion1.setType(DUMMYSTRUCTNAME.class);
    dmUnion1.read();
    dmUnion1.setType(POINT.class);
    dmUnion1.read();
    dmUnion1.setType(DUMMYSTRUCTNAME2.class); 
    dmUnion1.read();
} 
```

Reply:
@dbwiddis thanks.  I've taken a deep dive into `DEVMODE` and I'm starting to understand it's data.  I'm not entirely sure what happens when it's of unexpected length and that's something I need to do more research on.

For starters, I've placed JavaDoc comments into the Structure explaining exactly when a field is used, so this PR has a lot more DEVMODE information in it now.

I've also moved the fields around, I hope they're to your liking. :)

* `dmDriverExtra`: For example, the private driver data `dmDriverExtra` is meaningless to JNA, it can only ever be used by the driver manufacturer, since this format is intentionally proprietary.  I'm sure someone will eventually write a function to read this, but it may be easier to just extend the `DEVMODE` struct when that happens.   I'm fine leaving this alone, we just won't read the private data, as we have no place to put it. :)

* `dmSize`: I think this one is going to be our friend, but I haven't wrapped my head around how to use it.  What I've researched so far... 
   * `DUMMYUNIONNAME.DUMMYSTRUCTNAME` is NEVER used for Displays.  For this reason, I would like to rename it, but I don't know the ramifications of this.
   * `DUMMYUNIONNAME.DUMMYSTRUCTNAME2` equally is NEVER used for Printers.  For this reason, I would like to rename it as well, if that's a sane thing to do.
   * I have a suspicion that `dmSize` will give us some insight into which `DUMMYSTRUCT` will be provided, but I would love some help in this area if possible.  I can't find a bitwise flags (yet) that would just expose to use if this DEVMODE is of type PRINTER or DISPLAY, but I think (I hope) that this answers the original concerns.
   * One final option is I could make a skeleton DEVMODE interface and extend/implement it for DEVMODE_PRINTER, DEVMODE_DISPLAY, however I think this is a bit of overkill and would be less intuitive to the end-user.

Reply:
First ... not sure what you're doing with the `XP_OR_HIGHER` boolean.  If you do need it, use the Windows API `VersionHelpers.IsWindowsXPOrGreater()`.

To me it looks like `dmSize` would always be the same value which would match the JNA-calculated `size()` of this structure.   I don't think you can use it for choosing the union -- by definition the largest potential member is always used.  It appears `dmDriverExtra` can give a clue about a larger allocation you could use, but you have to first read the structure once to get that value, allocate something bigger, and then read it again to be able to use it.  Not sure that's a use case you need to handle, though.

It looks to me like you could use the `dmFields` value. The example in the API mentions the bit `dmOrientation` which is in the `DUMMYSTRUCTNAME` struct.   So if that bit is present, you'd have to read that member of the union.  Same for `dmPaperSize`, `dmPaperLength`, etc.   So you'd modify my code above to something like:
```java
@Override
public void read() {
    super.read();
    if (dmFields & (DM_ORIENTATION | DM_PAPERSIZE | DM_PAPERLENGTH 
            | DM_PAPERWIDTH | DM_SCALE | DM_COPIES | DM_DEFAULTSOURC
            | DM_PRINTQUALITY )) {
        dmUnion1.setType(DUMMYSTRUCTNAME.class);
        dmUnion1.read();
    }
    if (dmFields & (DM_POSITION)) {
        dmUnion1.setType(POINT.class);
        dmUnion1.read();
    }
    if (dmFields & (DM_POSITION | DM_DISPLAYORIENTATION | DM_DISPLAYFIXEDOUTPUT)) {
        dmUnion1.setType(DUMMYSTRUCTNAME2.class); 
        dmUnion1.read();
    }
} 
```

Reply:
Reading more it does look like there are different versions of the structure, thus the different version numbers and size fields available.  I'm guessing the newer version(s) added/would add fields at the end.   So if the size is smaller than `size()` you would only partially read the structure.  

So this falls into a typical Windows API situation of "call once to get the size and then call again with the properly sized buffer" pattern.